File: backend/src/modules/daily-diary/formatters.js
// Shared formatting utilities for daily-diary report files

export const parseJsonField = (val) => {
  if (val === null || val === undefined) return {};
  if (typeof val === 'string') {
    try { return JSON.parse(val); } catch { return {}; }
  }
  return val;
};

// Concatenated person details formatting formula:
// Name + "@ nick_name" + relation_prefix + relation_name + "R/O " + full_address + " Age- " + age + " yrs"
export const formatPerson = (name, age, fatherHusbandName, address, recordData = {}) => {
  let gender = null;
  let relationType = null;
  let nickname = null;
  
  if (recordData && typeof recordData === 'object') {
    const keys = Object.keys(recordData);
    let prefix = '';
    for (const k of keys) {
      if (k.endsWith('_relative_name') && recordData[k] === fatherHusbandName) {
        prefix = k.replace('_relative_name', '');
        break;
      }
      if (k.endsWith('_father_husband_name') && recordData[k] === fatherHusbandName) {
        prefix = k.replace('_father_husband_name', '');
        break;
      }
    }
    
    if (prefix) {
      gender = recordData[`${prefix}_gender`];
      relationType = recordData[`${prefix}_relation_type`] || recordData[`${prefix}_relationship`];
      nickname = recordData[`${prefix}_nickname`] || recordData[`${prefix}_nick_name`];
    }
    
    if (!gender) gender = recordData.gender || recordData.sex || recordData.arrested_gender || recordData.complainant_gender || recordData.accused_gender || recordData.deceased_gender;
    if (!relationType) relationType = recordData.relation_type || recordData.relationship || recordData.arrested_relation_type || recordData.complainant_relation_type || recordData.accused_relation_type || recordData.deceased_relation_type;
    if (!nickname) nickname = recordData.nickname || recordData.nick_name || recordData.arrested_nickname || recordData.complainant_nickname || recordData.accused_nickname;
  }
  
  const parts = [];
  
  // Name + nickname
  let nameStr = name || '';
  if (nickname && nameStr && !nameStr.includes(nickname)) {
    nameStr += ` @ ${nickname}`;
  }
  if (nameStr) parts.push(nameStr);
  
  // Relation
  let relPrefix = 'S/O';
  if (relationType) {
    const relLower = String(relationType).toLowerCase();
    if (relLower === 'husband') relPrefix = 'W/O';
    else if (relLower === 'father' || relLower === 'mother') {
      if (gender && String(gender).toLowerCase() === 'female') relPrefix = 'D/O';
      else relPrefix = 'S/O';
    } else relPrefix = 'C/O';
  } else {
    if (gender && String(gender).toLowerCase() === 'female') relPrefix = 'D/O';
  }
  
  if (fatherHusbandName) {
    parts.push(`${relPrefix} ${fatherHusbandName}`);
  }
  if (address) {
    parts.push(`R/O ${address}`);
  }
  if (age) {
    parts.push(`Age- ${age} yrs`);
  }
  return parts.join(' ') || '';
};

// "Name, S/O Father Name, Address" (no age — complainants, POs, traced persons)
export const formatPersonNoAge = (name, fatherHusbandName, address, recordData = {}) => {
  return formatPerson(name, null, fatherHusbandName, address, recordData);
};

// DD/MM/YYYY from any Date or ISO string
export const fmtDate = (dStr) => {
  if (!dStr) return '';
  if (dStr instanceof Date) {
    return `${String(dStr.getDate()).padStart(2, '0')}/${String(dStr.getMonth() + 1).padStart(2, '0')}/${dStr.getFullYear()}`;
  }
  const parts = String(dStr).split('T')[0].split('-');
  if (parts.length === 3) return `${parts[2]}/${parts[1]}/${parts[0]}`;
  return String(dStr);
};

// Column header labels shared across multiple sheets.
// Each report file can add its own labels — the index merges all of them.
export const SHARED_COLUMN_LABELS = {
  // Serial numbers
  sn: 'S.N.', sno: 'S.No.', s_no: 'S. No.', sr_no: 'Sr. No.', sr: 'Sr.',
  // Identifiers
  ps: 'Police Station',
  fir_no: 'FIR No.',
  efir_no: 'E-FIR No.',
  dd_nofir_no: 'DD No./FIR No.',
  firdd_no: 'FIR/DD No.',
  dd_no: 'DD No.',
  dd_date: 'DD Date',
  us: 'U/S',
  // IO fields
  io: 'Name of IO',
  name_of_io: 'Name of IO',
  io_name: 'IO Name',
  io_mobile_no: 'IO Mobile No.',
  rank_of_io: 'Rank of IO',
  mobile_no_of_io: 'Mobile No. of IO',
  // Combined person-detail columns
  complainant_details: 'Complainant (Name / S/O / Address)',
  arrested_details: 'Arrested Person (Name / Age / S/O / Address)',
  accused_details: 'Accused (Name / Age / S/O / Address)',
  po_details: 'PO Details (Name / S/O / Address)',
  deceased_details: 'Deceased (Name / Age / S/O / Address)',
  traced_person_details: 'Traced Person (Name / S/O / Address)',
  // Occurrence
  place_of_occurrence: 'Place of Occurrence',
  place_of_occurrence_1: 'Place of Occurrence (2)',
  time_of_occurrence: 'Time of Occurrence',
  date_of_occurrence: 'Date of Occurrence',
  // Common operational fields
  pcjcbail: 'PC/JC/Bail',
  recovery: 'Recovery',
  beat_no: 'Beat No.',
  stolen_items: 'Stolen Items',
  cause_of_death: 'Cause of Death',
  challan_untrace_cancel: 'Challan / Untrace / Cancel',
  name_of_operator_to_whom_mps: 'Name of Operator (MPS)',
  found_place: 'Found Place',
  found_date: 'Found Date',
  // Missing-person physical description
  upper_dress_color: 'Upper Dress Color',
  lower_dress_color: 'Lower Dress Color',
  // Arrest-specific
  prev_involvement_no_of_cases: 'Prev. Involvement (No. of Cases)',
  whether_accused_is_bc_or_not: 'Accused BC?',
  group_patrolling: 'Group Patrolling',
  cycle_patrolling: 'Cycle Patrolling',
  by_antisnatching_team: 'By Anti-Snatching Team',
  by_prahari: 'By Prahari',
  by_eyes_ears_scheme_members: 'By Eyes & Ears Scheme Members',
  // Summary-sheet columns
  pcr_call: 'PCR Call',
};

File: python_worker/formatters.py
from datetime import date as _date, datetime as _datetime


def fmt_date(d_str):
    """Return DD/MM/YYYY from an ISO date string, date, or datetime."""
    if not d_str:
        return ''
    if isinstance(d_str, (_date, _datetime)):
        return d_str.strftime('%d/%m/%Y')
    s = str(d_str).split('T')[0]
    parts = s.split('-')
    if len(parts) == 3:
        return f'{parts[2]}/{parts[1]}/{parts[0]}'
    return str(d_str)


def format_person(name, age, father_husband_name, address, record_data=None):
    """
    Build "Name @ Nickname S/O Father R/O Address Age- N yrs" string.
    Mirrors the JS formatPerson function exactly.
    """
    d = record_data or {}
    gender = None
    relation_type = None
    nickname = None

    if d:
        # Try to find which prefix owns father_husband_name
        prefix = ''
        for k, v in d.items():
            if k.endswith('_relative_name') and v == father_husband_name:
                prefix = k[:-len('_relative_name')]
                break
            if k.endswith('_father_husband_name') and v == father_husband_name:
                prefix = k[:-len('_father_husband_name')]
                break

        if prefix:
            gender = d.get(f'{prefix}_gender')
            relation_type = d.get(f'{prefix}_relation_type') or d.get(f'{prefix}_relationship')
            nickname = d.get(f'{prefix}_nickname') or d.get(f'{prefix}_nick_name')

        if not gender:
            gender = (d.get('gender') or d.get('sex') or d.get('arrested_gender')
                      or d.get('complainant_gender') or d.get('accused_gender') or d.get('deceased_gender'))
        if not relation_type:
            relation_type = (d.get('relation_type') or d.get('relationship')
                             or d.get('arrested_relation_type') or d.get('complainant_relation_type')
                             or d.get('accused_relation_type') or d.get('deceased_relation_type'))
        if not nickname:
            nickname = (d.get('nickname') or d.get('nick_name') or d.get('arrested_nickname')
                        or d.get('complainant_nickname') or d.get('accused_nickname'))

    parts = []

    name_str = name or ''
    if nickname and name_str and nickname not in name_str:
        name_str += f' @ {nickname}'
    if name_str:
        parts.append(name_str)

    rel_prefix = 'S/O'
    if relation_type:
        rel_lower = str(relation_type).lower()
        if rel_lower == 'husband':
            rel_prefix = 'W/O'
        elif rel_lower in ('father', 'mother'):
            rel_prefix = 'D/O' if gender and str(gender).lower() == 'female' else 'S/O'
        else:
            rel_prefix = 'C/O'
    else:
        if gender and str(gender).lower() == 'female':
            rel_prefix = 'D/O'

    if father_husband_name:
        parts.append(f'{rel_prefix} {father_husband_name}')
    if address:
        parts.append(f'R/O {address}')
    if age:
        parts.append(f'Age- {age} yrs')

    return ' '.join(parts)


def format_person_no_age(name, father_husband_name, address, record_data=None):
    return format_person(name, None, father_husband_name, address, record_data)
